Mathlib/MeasureTheory/Group/FundamentalDomain.lean

```diff
@@ -719,4 +727,281 @@ protected theorem fundamentalInterior : IsFundamentalDomain G (fundamentalInteri
+  MeasureTheory.HasFundamentalDomain
+
+/-- The `covolume` of an action of `G` on `α` which has a fundamental domain is the volume of some
+fundamental domain. -/
+@[to_additive addCovolume "The `addCovolume` of an action of `G` on `α` which has a fundamental
+domain is the volume of some fundamental domain."]
+def covolume (G α : Type*) [One G] [SMul G α] [MeasureSpace α]
+    [funDom : HasFundamentalDomain G α] : ℝ≥0∞ :=
+  volume funDom.has_fundamental_domain_characterization.choose
+
+variable [Group G] [MulAction G α] [MeasurableSpace G] [MeasureSpace α]
+
+/-- If there is a fundamental domain `s`, then `HasFundamentalDomain` holds. -/
+@[to_additive]
+lemma IsFundamentalDomain.hasFundamentalDomain {s : Set α} (fund_dom_s : IsFundamentalDomain G s) :
+    HasFundamentalDomain G α := ⟨⟨s, fund_dom_s⟩⟩
+
+/-- The `covolume` can be computed by taking the `volume` of any given fundamental domain `s`. -/
+@[to_additive]
+lemma IsFundamentalDomain.covolume_eq_volume [Countable G] [MeasurableSMul G α]
+    [SMulInvariantMeasure G α volume] {s : Set α} (fund_dom_s : IsFundamentalDomain G s) :
+    covolume (funDom := fund_dom_s.hasFundamentalDomain) = volume s := by
+  dsimp [covolume]
+  rw [fund_dom_s.measure_eq]
+  exact fund_dom_s.hasFundamentalDomain.has_fundamental_domain_characterization.choose_spec
+
+end HasFundamentalDomain
+
+/-! ## `QuotientVolumeEqVolume` typeclass
+
+This typeclass describes a situation in which a measure `μ` on `α ⧸ G` can be computed by
+taking the measure on `α` of the intersection of the pullback with a fundamental domain.
+
+It's curious that in measure theory, measures can be pushed forward, while in geometry, volumes can
+be pulled back. And yet here, we are describing a situation involving measures in a geometric way.
+
+Another viewpoint is that if a set is small enough to fit in a single fundamental domain, then its
+measure in `α` is the same as the `μ` measure of its pushforward in `α ⧸ G`.
+
+-/
+
+section QuotientVolumeEqVolume
+
+section additive
+
+variable [AddGroup G] [AddAction G α] [MeasureSpace α]
+
+local notation "α_mod_G" => AddAction.orbitRel G α
+
+local notation "π" => @Quotient.mk _ α_mod_G
+
+/-- A measure `μ` on the `AddQuotient` of `α` mod `G` satisfies `AddQuotientVolumeEqVolumePreimage`
+if: for any fundamental domain `t`, and any measurable subset `U` of the quotient,
+`μ U = volume ((π ⁻¹' U) ∩ t)`. -/
+class AddQuotientVolumeEqVolumePreimage (μ : Measure (Quotient α_mod_G)) : Prop where
+add_projection_respects_measure' : ∀ (t : Set α) (_ : IsAddFundamentalDomain G t)
+  (U : Set (Quotient α_mod_G)) (_ : MeasurableSet U),
+  μ U = volume (π ⁻¹' U ∩ t)
+
+end additive
+
+variable [Group G] [MulAction G α] [MeasureSpace α]
+
+local notation "α_mod_G" => MulAction.orbitRel G α
+
+local notation "π" => @Quotient.mk _ α_mod_G
+
+/-- A measure `μ` on the `Quotient` of `α` mod `G` satisfies `QuotientVolumeEqVolumePreimage` if:
```

Comment:
There is a strange asymmetry in this definition: you are requiring to have a MeasureSpace instance on the space alpha (so, with a canonical measure that you call volume), but on the other hand the measure on the quotient is an explicit measure. It would be more symmetric to require only a MeasurableSpace instance upstairs, and to let your class involve two measures, upstairs and downstairs. For this, you would need the class HasAddFundamentalDomain to also take an explicit measure, but you can let this measure be volume by default using a default argument as you are doing elsewhere.

Reply:
We've changed `[MeasureSpace alpha]` to `[MeasurableSpace]` and added a measure nu on alpha. This means that a few instances can no longer be obtained just from typeclass inference, so we had to make them lemmas.